src/main/java/jenkins/plugins/git/GitToolChooser.java

```diff
@@ -0,0 +1,181 @@
+package jenkins.plugins.git;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.EnvVars;
+import hudson.ExtensionList;
+import hudson.ExtensionPoint;
+import hudson.model.TaskListener;
+import hudson.plugins.git.GitTool;
+import hudson.plugins.git.util.GitUtils;
+import jenkins.model.Jenkins;
+import org.apache.commons.io.FileUtils;
+import org.jenkinsci.plugins.gitclient.Git;
+import org.jenkinsci.plugins.gitclient.GitClient;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+/**
+ * A class which allows Git Plugin to choose a git implementation by estimating the size of a repository from a distance
+ * without requiring a local checkout.
+ */
+public class GitToolChooser {
+
+    private long sizeOfRepo = 0L;
+    private String implementation;
+    private String gitTool;
+    /**
+     * Size to switch implementation in KiB
+     */
+    public static final int SIZE_TO_SWITCH = 5000;
+
+    /**
+     * Instantiate class using {@link AbstractGitSCMSource}. It looks for a cached .git directory first, calculates the
+     * size if it is found else checks if the extension point has been implemented and asks for the size.
+     * @param source the {@link AbstractGitSCMSource}
+     * @throws IOException
+     * @throws InterruptedException
+     */
+    public GitToolChooser(@NonNull AbstractGitSCMSource source) throws IOException, InterruptedException {
+        boolean useCache;
+        boolean useAPI = false;
+
+        implementation = determineSwitchOnSize(sizeOfRepo);
+        useCache = setSizeFromCache(source);
+
+        if (useCache) {
+            implementation = determineSwitchOnSize(sizeOfRepo);
+        } else {
+            useAPI = setSizeFromAPI(source.getRemote());
+        }
+
+        if (useAPI) {
+            implementation = determineSwitchOnSize(sizeOfRepo);
+        }
+        determineGitTool(implementation);
+    }
```

Comment:
will try to explain better my outdated comment, because it was so clear in my mind that probably I lack an explanation for you. Let's go step by step:
1. `useCache` and `useAPI` are initialized to `false`
1. Check `setSizeFromCache`
   1. if `true`, then use `determineSwitchOnSize` and `useAPI` will be always `false`
   1. If `false` then use `setSizeFromAPI` and `useAPI` is updated
1. The only way that `determineSwitchOnSize` can be executed is if `useAPI` is false and `setSizeFromAPI` returns `true`, so
```suggestion
    public GitToolChooser(@NonNull AbstractGitSCMSource source) throws IOException, InterruptedException {
        boolean useCache;
        boolean useAPI = false;

        implementation = determineSwitchOnSize(sizeOfRepo);
        useCache = setSizeFromCache(source);

        if (useCache) {
            implementation = determineSwitchOnSize(sizeOfRepo);
        } else {
            useAPI = setSizeFromAPI(source.getRemote());
            if (useAPI) {
                implementation = determineSwitchOnSize(sizeOfRepo);
            }
        }
        determineGitTool(implementation);
    }
```
As I said on another outdated comment you can find duplicated code between the two of the constructors:
```
        boolean useAPI = setSizeFromAPI(remoteName);
        if (useAPI) {
            implementation = determineSwitchOnSize(sizeOfRepo);
        }
``` 
so then:
```suggestion
    public GitToolChooser(@NonNull AbstractGitSCMSource source) throws IOException, InterruptedException {
        implementation = "NONE";
        if (setSizeFromCache(source)) {
            implementation = determineSwitchOnSize(sizeOfRepo);
        } else {
            useAPI()
        }
    }
```

where the new `useAPI()` method would be
```
    private useAPI() {
        if (setSizeFromAPI(remoteName)) {
            implementation = determineSwitchOnSize(sizeOfRepo);
        }
    }
```
This suggestion is not changing your logic, which is correct, it's trying to optimize the code preventing the duplicity.

Reply:
@fcojfernandez Thanks for explaining the duplication so clearly, in my mind, I am always thinking about readability of the code and in that process I added duplicate code. I have tried to fix in the subsequent commits, please review them whenever you get the time!